packages/cardano-graphql/test/WalletProvider/CardanoGraphQLWalletProvider.test.ts

```diff
@@ -0,0 +1,114 @@
+      });
+    });
+
+    it('throws ProviderError{NotFound} on empty response', async () => {
+      sdk.Tip.mockResolvedValueOnce({});
+      await expect(provider.ledgerTip()).rejects.toThrow(ProviderFailure.NotFound);
+      sdk.Tip.mockResolvedValueOnce({ queryBlock: [] });
+      await expect(provider.ledgerTip()).rejects.toThrow(ProviderFailure.NotFound);
+    });
+
+    it('throws ProviderError{InvalidResponse} if provider returns multiple tips', async () => {
+      sdk.Tip.mockResolvedValueOnce({ queryBlock: [tip, tip] });
+      await expect(provider.ledgerTip()).rejects.toThrow(ProviderFailure.InvalidResponse);
+    });
+  });
+
+  describe('queryBlocksByHashes', () => {
+    const block = {
+      blockNo: 1,
+      confirmations: 4,
+      epoch: { number: 5 },
+      fees: '123',
+      hash: '6804edf9712d2b619edb6ac86861fe93a730693183a262b165fcc1ba1bc99cad',
+      nextBlock: { hash: '6804edf9712d2b619edb6ac86861fe93a730693183a262b165fcc1ba1bc99cae' },
+      previousBlock: { hash: '6804edf9712d2b619edb6ac86861fe93a730693183a262b165fcc1ba1bc99caa' },
+      size: 6,
+      slot: { date: '2019-10-12T07:20:50.52Z', number: 2, slotInEpoch: 3 },
+      slotLeader: { id: 'pool1zuevzm3xlrhmwjw87ec38mzs02tlkwec9wxpgafcaykmwg7efhh' },
+      totalOutput: '700',
+      transactionsAggregate: {
+        count: 3
+      },
+      vrf: 'vrf_vk19j362pkr4t9y0m3qxgmrv0365vd7c4ze03ny4jh84q8agjy4ep4s99zvg8'
+    } as NonNullable<NonNullable<BlocksByHashesQuery['queryBlock']>[0]>;
+    const blockHash = Cardano.BlockId(block.hash);
+
+    it('makes a graphql query and coerces result to core types', async () => {
+      sdk.BlocksByHashes.mockResolvedValueOnce({
+        queryBlock: [block, { ...block, hash: '6804edf9712d2b619edb6ac86861fe93a730693183a262b165fcc1ba1bc99caa' }]
+      });
+      const blocks = await provider.queryBlocksByHashes([
+        blockHash,
+        Cardano.BlockId('6804edf9712d2b619edb6ac86861fe93a730693183a262b165fcc1ba1bc99caa')
+      ]);
+      expect(blocks).toHaveLength(2);
+      expect(blocks[0]).toEqual({
+        confirmations: block.confirmations,
+        date: new Date(block.slot.date),
+        epoch: block.epoch.number,
+        epochSlot: block.slot.slotInEpoch,
+        fees: BigInt(block.fees),
+        header: {
+          blockNo: block.blockNo,
+          hash: block.hash,
+          slot: block.slot.number
+        },
+        nextBlock: block.nextBlock.hash,
+        previousBlock: block.previousBlock.hash,
+        size: block.size,
+        slotLeader: block.slotLeader.id,
+        totalOutput: BigInt(block.totalOutput),
+        txCount: block.transactionsAggregate!.count,
+        vrf: block.vrf
+      });
+    });
+
+    it('returns an empty array on undefined response', async () => {
+      sdk.BlocksByHashes.mockResolvedValueOnce({});
+      expect(await provider.queryBlocksByHashes([blockHash])).toEqual([]);
+    });
+
+    // Review: should this be interpreted as invalid response instead?
```

Comment:
Not sure, and will depend on how aggregates are handled by Dgraph. Can probably leave as-is and we can refine